tendermint/src/lite/public.rs

```diff
@@ -0,0 +1,170 @@
+        let trusted_vals = trusted_state.validators();
+        verify_commit_trusting(trusted_vals, untrusted_commit, trust_threshold)?;
+    }
+
+    // verify the untrusted commit
+    verify_commit_full(untrusted_vals, untrusted_sh.commit())
+}
+
+/// Attempt to update the store to the given untrusted header.
+/// Ensures our last trusted header hasn't expired yet, and that
+/// the untrusted header can be verified using only our latest trusted
+/// state from the store.
+/// This function is primarily for use by IBC handlers.
+pub fn verify_and_update_single<TS, SH, C, L, S>(
+    untrusted_sh: &SH,
+    untrusted_vals: &C::ValidatorSet,
+    untrusted_next_vals: &C::ValidatorSet,
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
+    SH: SignedHeader<Commit = C>,
+    C: Commit,
+    L: TrustThreshold,
+    S: Store<TrustedState = TS>,
+{
+    // fetch the latest state and ensure it hasn't expired
+    let trusted_state = store.get(Height::from(0))?;
+    let trusted_sh = trusted_state.last_header();
+    is_within_trust_period(trusted_sh.header(), trusting_period, now)?;
+
+    verify_single(
+        trusted_state,
+        untrusted_sh,
+        untrusted_vals,
+        untrusted_next_vals,
+        trust_threshold,
+    )?;
+
+    // the untrusted header is now trusted. update the store
+    let new_trusted_state = TS::new(untrusted_sh, untrusted_next_vals);
+    store.add(&new_trusted_state)
+}
+
+/// Attempt to update the store to the given untrusted height.
+/// This function is recursive: it uses a bisection algorithm
+/// to request data for intermediate heights as necessary.
+/// Ensures our last trusted header hasn't expired yet, and that
+/// data from the untrusted height can be verified, possibly using
+/// data from intermediate heights.
+/// This function is primarily for use by a light node.
+pub fn verify_and_update_bisection<TS, SH, C, L, R, S>(
+    untrusted_height: Height,
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
+    SH: SignedHeader<Commit = C>,
+    C: Commit,
+    L: TrustThreshold,
+    R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
+    S: Store<TrustedState = TS>,
+{
+    // Fetch the latest state and ensure it hasn't expired.
+    // Note when calling itself recursively, this check is redundant,
+    // but leaving it keeps things simple ...
+    let trusted_state = store.get(Height::from(0))?;
+    let trusted_sh = trusted_state.last_header();
+    is_within_trust_period(trusted_sh.header(), trusting_period, now)?;
+
+    // fetch the header and vals for the new height
+    let untrusted_sh = &req.signed_header(untrusted_height)?;
```

Comment:
Where do we check if downloaded header is within trusted period?

Reply:
We don't. The thinking is that our local `now` is close to the latest blockchain time, and if `now` is within the trusted period, so is the blockchain time. Also once the new height is stored, it will be checked for expiry the next time we try to update.

That said, it's probably not a bad idea to check that its within trusting period from the last trusted state, or, what should be equivalent, that its less than `now + X`, where X is our allowed variance between our local time and the blockchains latest bft time.

Reply:
I added a comment about this in 60c8b16059c21ecbd72319ded47dcb68b2428cb0